key_management/src/wallet.rs

```diff
@@ -0,0 +1,181 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use super::errors::Error;
+use super::{sigs, KeyInfo, KeyStore};
+use address::Address;
+use crypto::{Signature, SignatureType};
+use std::collections::HashMap;
+use std::str::FromStr;
+
+/// A Key, this contains a key_info, address, and public_key which holds the key type and private key
+#[derive(Clone, PartialEq, Debug, Eq)]
+pub struct Key {
+    key_info: KeyInfo,
+    // Vec<u8> will be used because Eq has not been implemented for PublicKey type
+    public_key: Vec<u8>,
+    address: Address,
+}
+
+impl Key {
+    /// Return a new Key given key_info
+    pub fn new(key_info: &KeyInfo) -> Result<Key, Error> {
+        let public_key =
+            sigs::to_public(act_sig_type(key_info.key_type()), key_info.private_key())?;
+        let address = sigs::new_address(act_sig_type(key_info.key_type()), public_key.clone())?;
+        Ok(Key {
+            key_info: key_info.clone(),
+            public_key,
+            address,
+        })
+    }
+}
+
+/// This is a Wallet, it contains 2 HashMaps:
+/// - keys which is a HashMap of Keys resolved by their Address
+/// - keystore which is a HashMap of KeyInfos resolved by their Address
+#[derive(Clone, PartialEq, Debug, Eq)]
+pub struct Wallet {
+    keys: HashMap<Address, Key>,
+    keystore: KeyStore,
+}
+
+impl Wallet {
+    /// Return a new Wallet with a given KeyStore
+    pub fn new(keystore: KeyStore) -> Self {
+        Wallet {
+            keys: HashMap::new(),
+            keystore,
+        }
+    }
+
+    /// Return a new Wallet with Keys contructed by a given list of Keys
+    pub fn new_from_keys(key_vec: Vec<Key>) -> Self {
+        let mut keys: HashMap<Address, Key> = HashMap::new();
+        for item in key_vec.clone() {
+            keys.insert(item.address.clone(), item);
+        }
+        Wallet {
+            keys,
+            keystore: KeyStore::new(),
+        }
+    }
+
+    /// Return the Key that is resolved by a given Address, return Error otherwise
+    pub fn find_key(&mut self, addr: &Address) -> Result<Key, Error> {
+        let key = self.keys.get(&addr);
+        if let Some(k) = key {
+            return Ok(k.clone());
+        }
+        let mut owned_string = "wallet-".to_owned();
+        owned_string.push_str(addr.to_string().as_ref());
+        let key_info = self.keystore.get(&owned_string)?;
+        let new_key = Key::new(key_info)?;
+        self.keys.insert(addr.clone(), new_key.clone());
```

Comment:
Doesn't matter tho, can leave this
```suggestion
        self.keys.insert(*addr, new_key.clone());
```